rasterio/_io.pyx

```diff
@@ -643,45 +643,58 @@ cdef class DatasetReaderBase(DatasetBase):
             else:
                 nodataval = ndv
 
-            # mask_vrt_doc is bytes, the dataset ctor needs a str.
-            # Allowing it to take bytes is a TODO.
             vrt_doc = _boundless_vrt_doc(
-                self, nodata=nodataval, background=nodataval,
+                self,
+                nodata=nodataval,
+                background=nodataval,
                 width=max(self.width, window.width) + 1,
                 height=max(self.height, window.height) + 1,
                 transform=self.window_transform(window)
-            ).decode("utf-8")
-
-            vrt_kwds = {'driver': 'VRT'}
-            with DatasetReaderBase(_UnparsedPath(vrt_doc), **vrt_kwds) as vrt:
+            )
 
-                out = vrt._read(
-                    indexes, out, Window(0, 0, window.width, window.height),
-                    None, resampling=resampling)
+            with VirtualDataset.fromstring(vrt_doc) as vrt:
+                with vrt.open() as dataset:
```

Comment:
The old lines 655-656 were kinda gross. The new lines 655-656 are better. Fewer implementation details like `_UnparsedPath`.

I don't know what the arguments for `VirtualDataset` are yet, but I'm sure that we want a `fromstring` classmethod. If this VirtualDataset idea feels good, I'll replace _boundless_vrt_doc with a function that returns VirtualDataset.

Note: the reason why VirtualDataset is used as a context manager is that GDALOpenEx has a filename length limit. For complex VRTs, we need a temp file as a home for the XML.